File: 10-Recursion-and-Backtracking/python/04_m_coloring_problem.py
class Solution:
    def graphColoring(self, edges: List[List[int]], m: int, n: int) -> bool:
        adj = [[] for _ in range(n)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        color = [0] * n

        def is_safe(node, c):
            for neighbor in adj[node]:
                if color[neighbor] == c:
                    return False
            return True

        def solve(node):
            if node == n:
                return True

            for c in range(1, m + 1):
                if is_safe(node, c):
                    color[node] = c
                    if solve(node + 1):
                        return True
                    color[node] = 0
            return False

        return solve(0)

File: 10-Recursion-and-Backtracking/java/04-MColoringProblem.java
class Solution {
    boolean graphColoring(int[][] edges, int m, int n) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());
        for (int[] edge : edges) {
            int u = edge[0], v = edge[1];
            adj.get(u).add(v);
            adj.get(v).add(u);
        }

        int[] color = new int[n];
        return solve(0, n, m, adj, color);
    }

    private boolean solve(int node, int n, int m, List<List<Integer>> adj, int[] color) {
        if (node == n) return true;

        for (int c = 1; c <= m; c++) {
            if (isSafe(node, c, adj, color)) {
                color[node] = c;
                if (solve(node + 1, n, m, adj, color)) return true;
                color[node] = 0;
            }
        }
        return false;
    }

    private boolean isSafe(int node, int c, List<List<Integer>> adj, int[] color) {
        for (int neighbor : adj.get(node)) {
            if (color[neighbor] == c) return false;
        }
        return true;
    }
}
